source/common/thread_local/thread_local_impl.cc

```diff
@@ -168,8 +168,11 @@ void InstanceImpl::runOnAllThreads(Event::PostCb cb, Event::PostCb all_threads_c
 
   Event::PostCbSharedPtr cb_guard(new Event::PostCb(cb),
                                   [this, all_threads_complete_cb](Event::PostCb* cb) {
-                                    main_thread_dispatcher_->post(all_threads_complete_cb);
+                                    // We need to delete the cb before posting
+                                    // to the main thread to avoid a race that
+                                    // can occur if the completion callback deletes the slot.
```

Comment:
To help with review can you add a bit more comments either here in the comment thread or to the code comment about what the actual race condition is? It's not immediately clear to me from reading.

Reply:
The `cb_guard` is a shared_ptr to a `Event::Post(cb)`, where that `cb` is the result of calling `SlotImpl::dataCallback` which returns a lambda that captures the SlotImpl’s `still_alive_guard_` via a weak_ptr. The `cb_guard` uses the shared_ptr custom deleter to post the completion callback to the main thread, and will be invoked by the last dtor of the last shared_ptr to `cb_guard`. 

I think the following is the race:

1. The worker thread runs the dtor of the cb guard
2. The main thread gets the completion callback posted to it, executes the callback to delete the slot, which deletes the slots shared_ptr to `still_alive_guard_`
3. The worker thread continues the dtor of the cb guard and deletes the `cb` generated by `SlotImpl::dataCallback` which is a lambda with a weak_ptr to `still_alive_guard_`, seeing as it’s the last entity pointing to the control block for `still_alive_guard_`  it goes to free the memory.

TSAN puts it as a WAW or RAW.

TSAN dump:

Worker thread:
```
  Write of size 8 at 0x7b080002ab68 by thread T1:
    #0 free <null> (thread_local_impl_test+0x29ac218)
    #1 std::__1::_DeallocateCaller::__do_call(void*) <null> (thread_local_impl_test+0x2a509c8)
    #2 std::__1::_DeallocateCaller::__do_deallocate_handle_size(void*, unsigned long) <null> (thread_local_impl_test+0x2a50944)
    #3 std::__1::_DeallocateCaller::__do_deallocate_handle_size_align(void*, unsigned long, unsigned long) <null> (thread_local_impl_test+0x2a5086d)
    #4 std::__1::__libcpp_deallocate(void*, unsigned long, unsigned long) <null> (thread_local_impl_test+0x2a507c8)
    #5 std::__1::allocator<std::__1::__shared_ptr_emplace<bool, std::__1::allocator<bool> > >::deallocate(std::__1::__shared_ptr_emplace<bool, std::__1::allocator<bool> >*, unsigned long) <null> (thread_local_impl_test+0x2b3ac1d)
    #6 std::__1::__shared_ptr_emplace<bool, std::__1::allocator<bool> >::__on_zero_shared_weak() <null> (thread_local_impl_test+0x2b3a695)
    #7 std::__1::weak_ptr<bool>::~weak_ptr() <null> (thread_local_impl_test+0x2b33417)
    #8 Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1::~$_1() <null> (thread_local_impl_test+0x2b1d822)
    #9 std::__1::__compressed_pair_elem<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1, 0, false>::~__compressed_pair_elem() <null> (thread_local_impl_test+0x2b269f8)
    #10 std::__1::__compressed_pair<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1, std::__1::allocator<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1> >::~__compressed_pair() <null> (thread_local_impl_test+0x2b26cab)
    #11 std::__1::__function::__alloc_func<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1, std::__1::allocator<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1>, void ()>::destroy() <null> (thread_local_impl_test+0x2b27d98)
    #12 std::__1::__function::__func<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1, std::__1::allocator<Envoy::ThreadLocal::InstanceImpl::SlotImpl::dataCallback(std::__1::function<void (std::__1::shared_ptr<Envoy::ThreadLocal::ThreadLocalObject>)> const&)::$_1>, void ()>::destroy_deallocate() <null> (thread_local_impl_test+0x2b2634e)
    #13 std::__1::__function::__value_func<void ()>::~__value_func() <null> (thread_local_impl_test+0x2a853cc)
    #14 std::__1::function<void ()>::~function() <null> (thread_local_impl_test+0x2a4cc18)
    #15 Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_5::operator()(std::__1::function<void ()>*) const <null> (thread_local_impl_test+0x2b2f35f)
    #16 std::__1::__shared_ptr_pointer<std::__1::function<void ()>*, Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_5, std::__1::allocator<std::__1::function<void ()> > >::__on_zero_shared() <null> (thread_local_impl_test+0x2b2f6d3)
    #17 std::__1::__shared_count::__release_shared() <null> (thread_local_impl_test+0x2a5500c)
    #18 std::__1::__shared_weak_count::__release_shared() <null> (thread_local_impl_test+0x2a54f82)
    #19 std::__1::shared_ptr<std::__1::function<void ()> >::~shared_ptr() <null> (thread_local_impl_test+0x2b343e7)
    #20 Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_6::~$_6() <null> (thread_local_impl_test+0x2b219f8)
    #21 std::__1::__compressed_pair_elem<Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_6, 0, false>::~__compressed_pair_elem() <null> (thread_local_impl_test+0x2b30bd8)
    #22 std::__1::__compressed_pair<Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_6, std::__1::allocator<Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_6> >::~__compressed_pair() <null> (thread_local_impl_test+0x2b30deb)
    #23 std::__1::__function::__alloc_func<Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_6, std::__1::allocator<Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_6>, void ()>::destroy() <null> (thread_local_impl_test+0x2b320c8)
    #24 std::__1::__function::__func<Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_6, std::__1::allocator<Envoy::ThreadLocal::InstanceImpl::runOnAllThreads(std::__1::function<void ()>, std::__1::function<void ()>)::$_6>, void ()>::destroy() <null> (thread_local_impl_test+0x2b30512)
    #25 std::__1::__function::__value_func<void ()>::~__value_func() <null> (thread_local_impl_test+0x2a85370)
    #26 std::__1::function<void ()>::~function() <null> (thread_local_impl_test+0x2a4cc18)
    #27 void std::__1::allocator_traits<std::__1::allocator<std::__1::__list_node<std::__1::function<void ()>, void*> > >::__destroy<std::__1::function<void ()> >(std::__1::integral_constant<bool, false>, std::__1::allocator<std::__1::__list_node<std::__1::function<void ()>, void*> >&, std::__1::function<void ()>*) <null> (thread_local_impl_test+0x2a7f804)
    #28 void std::__1::allocator_traits<std::__1::allocator<std::__1::__list_node<std::__1::function<void ()>, void*> > >::destroy<std::__1::function<void ()> >(std::__1::allocator<std::__1::__list_node<std::__1::function<void ()>, void*> >&, std::__1::function<void ()>*) <null> (thread_local_impl_test+0x2a7f718)
    #29 std::__1::list<std::__1::function<void ()>, std::__1::allocator<std::__1::function<void ()> > >::pop_front() <null> (thread_local_impl_test+0x2a7f97a)
    #30 Envoy::Event::DispatcherImpl::runPostCallbacks() <null> (thread_local_impl_test+0x527e0fa)
```


Main thread:
```
  Previous atomic write of size 8 at 0x7b080002ab68 by main thread:
    #0 __tsan_atomic64_fetch_add <null> (thread_local_impl_test+0x29f9351)
    #1 long std::__1::__libcpp_atomic_refcount_decrement<long>(long&) <null> (thread_local_impl_test+0x2a55080)
    #2 std::__1::__shared_count::__release_shared() <null> (thread_local_impl_test+0x2a54fe9)
    #3 std::__1::__shared_weak_count::__release_shared() <null> (thread_local_impl_test+0x2a54f82)
    #4 std::__1::shared_ptr<bool>::~shared_ptr() <null> (thread_local_impl_test+0x2b35707)
    #5 Envoy::ThreadLocal::InstanceImpl::SlotImpl::~SlotImpl() <null> (thread_local_impl_test+0x2b34977)
    #6 Envoy::ThreadLocal::InstanceImpl::SlotImpl::~SlotImpl() <null> (thread_local_impl_test+0x2b349ef)
    #7 std::__1::default_delete<Envoy::ThreadLocal::Slot>::operator()(Envoy::ThreadLocal::Slot*) const <null> (thread_local_impl_test+0x2a7fd56)
    #8 std::__1::unique_ptr<Envoy::ThreadLocal::Slot, std::__1::default_delete<Envoy::ThreadLocal::Slot> >::reset(Envoy::ThreadLocal::Slot*) <null> (thread_local_impl_test+0x2a7fc70)
    #9 std::__1::unique_ptr<Envoy::ThreadLocal::Slot, std::__1::default_delete<Envoy::ThreadLocal::Slot> >::~unique_ptr() <null> (thread_local_impl_test+0x2a7fbbc)
    #10 Envoy::ThreadLocal::TypedSlot<Envoy::ThreadLocal::ThreadLocalObject>::~TypedSlot() <null> (thread_local_impl_test+0x2a4cde8)
    #11 std::__1::default_delete<Envoy::ThreadLocal::TypedSlot<Envoy::ThreadLocal::ThreadLocalObject> >::operator()(Envoy::ThreadLocal::TypedSlot<Envoy::ThreadLocal::ThreadLocalObject>*) const <null> (thread_local_impl_test+0x2b06686)
    #12 std::__1::unique_ptr<Envoy::ThreadLocal::TypedSlot<Envoy::ThreadLocal::ThreadLocalObject>, std::__1::default_delete<Envoy::ThreadLocal::TypedSlot<Envoy::ThreadLocal::ThreadLocalObject> > >::reset(Envoy::ThreadLocal::TypedSlot<Envoy::ThreadLocal::ThreadLocalObject>*) <null> (thread_local_impl_test+0x2a4b7d0)
    #13 Envoy::ThreadLocal::ThreadLocalInstanceImplDispatcherTest_RaceConditionIfCompletionCallback_Test::TestBody()::$_13::operator()() const::'lambda'()::operator()() const <null> (thread_local_impl_test+0x2a4843f)
```
By deleting the callback in the cb_guard before posting the completion callback we ensure there's no weak_ptr reference to the slot's `still_alive_guard_` which the main thread will delete in the completion callback.

Reply:
Hmm, I've seen traces like this before, and in fact I think I saw traces like this when I was implementing this code. Isn't this just due to not using the TSAN instrumented version of libc as per https://github.com/envoyproxy/envoy/tree/main/bazel#sanitizers? 

(This trace would indicate to me an issue in libc and not Envoy, AIUI this trace should be safe.)

/wait-any